### PR TITLE
[master] ZIL-5146: only allow lookup & seedprv node types to use telemetry 

### PR DESCRIPTION
**Pull request by yaron-zilliqa**
## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [ ] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libMetrics/Metrics.cpp [hunk 1]
```diff
@@ -39,31 +39,41 @@
...
+  // c-tor
+  zil::metrics::Filter::GetInstance().init();
+}
 
-void Metrics::Init() {
-  std::string cmp(METRIC_ZILLIQA_PROVIDER);
+void Metrics::Initialize(std::string_view identity /*= {}*/,
+                         std::string provider /* = METRIC_ZILLIQA_PROVIDER*/) {
+  static std::array<std::string_view, 2> allowedIdentities = {"seedprv",
+                                                              "lookup"};
```
Comment (JamesHinshelwood) [inline src/libMetrics/Metrics.cpp:57]: Hmm, do we want metrics, tracing, etc. for normals, dsguards, seedpubs as well? I would assume so?

Comment (yaron-zilliqa) [reply src/libMetrics/Metrics.cpp:57]: We do, but the identity I believe should be also lookup. Unless the init.py in testnet script is incorrect:

`parser.add_argument('--type', choices=['normal', 'lookup', 'new', 'seedprv'], help='node type (normal, lookup, new, seedprv)')
`

Comment (yaron-zilliqa) [reply src/libMetrics/Metrics.cpp:57]: I think you're right and these aren't actually the only options. I'll add the others. Wouldn't hurt anyway.

Comment (JamesHinshelwood) [reply src/libMetrics/Metrics.cpp:57]: I did a quick grep of `testnet`. I believe `--type` is only set for the join pod. For the others, it is inferred from the pod name by splitting on `-`. So if I understand correctly, the other nodes will have type "dsguard", "normal", etc.

Comment (JamesHinshelwood) [reply src/libMetrics/Metrics.cpp:57]: I guess the important question is what is the `--type` of external miner nodes? Is it set in `launch_docker.sh`?

File: src/libMetrics/Logging.cpp [hunk 2]
```diff
@@ -142,25 +142,36 @@ std::shared_ptr<logs::LoggerProvider> InitNoop() {
...
-Logging::Logging() {
-  std::string cmp(LOGGING_ZILLIQA_PROVIDER);
+void Logging::Initialize(std::string_view identity /*= {}*/,
+                         std::string provider /*= LOGGING_ZILLIQA_PROVIDER*/) {
+  static std::array<std::string_view, 4> allowedIdentities = {
+      "seedprv", "seedpub", "dsguard", "lookup"};
+  if (std::none_of(std::begin(allowedIdentities), std::end(allowedIdentities),
+                   [identity](const auto& allowedIdentity) {
+                     return identity.find(allowedIdentity) != std::string::npos;
+                   })) {
```
Comment (rrw-zilliqa) [inline src/libMetrics/Logging.cpp:152]: Works, but a bit convoluted? Might have been better to write an aux function (or at least a comment):

```
if (!std::any_of(std::begin(allowedIdentities), std::end(allowedIdentities), 
   /* Is this allowed? */
    [identity](const auto& allowedIdentity) { return identity.find(allowedIdentity) != std::string::npos } })) { 
     .. 
   }
```
?

File: src/libMetrics/Logging.cpp [hunk 3]
```diff
@@ -142,25 +142,36 @@ std::shared_ptr<logs::LoggerProvider> InitNoop() {
...
+    provider = "NONE";
+  }
+
+  transform(provider.begin(), provider.end(), provider.begin(), ::tolower);
```
Comment (rrw-zilliqa) [inline src/libMetrics/Logging.cpp:156]: Worth a utility function? (or using the boost alg?) - this is really nitpicky!

File: src/libMetrics/Metrics.cpp [hunk 4]
```diff
@@ -39,31 +39,41 @@
...
+  // c-tor
+  zil::metrics::Filter::GetInstance().init();
+}
 
-void Metrics::Init() {
-  std::string cmp(METRIC_ZILLIQA_PROVIDER);
+void Metrics::Initialize(std::string_view identity /*= {}*/,
+                         std::string provider /* = METRIC_ZILLIQA_PROVIDER*/) {
+  static std::array<std::string_view, 4> allowedIdentities = {
+      "seedprv", "seedpub", "dsguard", "lookup"};
```
Comment (rrw-zilliqa) [inline src/libMetrics/Metrics.cpp:57]: Use a singleton rather than duplicating the list?

**Review by rrw-zilliqa — APPROVED** [review]
Looks good to me :-)

File: src/libMetrics/Logging.cpp (hunk 2 above)
Comment (yaron-zilliqa) [reply src/libMetrics/Logging.cpp]: As a matter of fact that's how I initially wrote it and thought that was a bit less clear :D .... at any rate - I've moved this to an aux function and reused in Metrics.cpp & Tracing.cpp as well.

File: src/libMetrics/Logging.cpp (hunk 3 above)
Comment (yaron-zilliqa) [reply src/libMetrics/Logging.cpp]: Yeah, changed to use the boost::algorithm::to_lower.

File: src/libMetrics/Metrics.cpp (hunk 4 above)
Comment (yaron-zilliqa) [reply src/libMetrics/Metrics.cpp]: Moved all this to a separate aux function so it's not repeated. (regardless, I hate singletons and don't really use them).

File: src/libMetrics/Common.h [hunk 5]
```diff
@@ -32,4 +34,15 @@ const std::string METRIC_SCHEMA{"https://opentelemetry.io/schemas/1.2.0"};
 
 constexpr double METRICS_VERSION{8.6};
 
+template <typename StringT>
+bool IsObservabilityAllowed(const StringT& identity) {
+  static std::array<std::string_view, 4> allowedIdentities = {
+      "seedprv", "seedpub", "dsguard", "lookup"};
```
Comment (JamesHinshelwood) [inline src/libMetrics/Common.h:40]: Should "normal" be here too?

**Pull request merged**